nsc-events-nestjs_18_619(nextjs)_add-dropdown-box-to-user-management-backend

Asfand00 opened the pull request:
Helps Resolve issue (#619) on the nextjs side 
PR link for that: https://github.com/SeattleColleges/nsc-events-nextjs/pull/627  

This PR is for adding some necessary changes to make sure the role filter dropdown box works on the frontend side. 

With help from @bennettsf, the main files that were modified are `user.controller.ts` and `user.service.ts` to ensure that the backend is receiving the update from the frontend so that the role filter works correctly. 

This branch should be used in conjunction to the frontend branch that is listed in the PR link above

To test this branch on its own for functionality:
- first clone then run `npm install`
- ensure backend local database is setup with several users of different roles (MongoDB)
- then run `npm run start:dev` to start backend
- By using Postman make a GET request at http://localhost:3001/api/users/search?page=1&role=admin 
- you can change the `role=` part to a different role (user, creator). 

Here is my test of using Postman of that same link: 
![image](https://github.com/user-attachments/assets/fc996196-9914-45a1-9072-7a25e634d29f)

Update: fixed the pagination issue and removed the @Roles/@Guard comments as per requested by @bennettsf .

bennettsf reviewed (changes requested):
Make sure to uncomment the @Roles/@UseGuards code to make sure admins only have access to the endpoint. Also modify L55 of the service file to `const currentPage = Number(page) || 1;` to fix the bug we were running into with pagination. 

Otherwise, everything seems to be working properly. Once you push those changes through, I'll clean up the return/input models so that the code is more organized/readable.

CourtneyHoppus reviewed (changes requested):
I tested this in Postman, with admin, user, and creator each for a search parameter:
<img width="721" alt="Screenshot 2025-03-03 at 10 43 46 AM" src="https://github.com/user-attachments/assets/4462da43-9671-4b28-9416-cba4797149f0" />
<img width="721" alt="Screenshot 2025-03-03 at 10 44 00 AM" src="https://github.com/user-attachments/assets/d2e1b944-d7a1-4211-a394-c00cab6b15e0" />
<img width="721" alt="Screenshot 2025-03-03 at 10 44 08 AM" src="https://github.com/user-attachments/assets/929e3472-4a6c-40e9-b4f7-4da146d2f9c5" />


As @bennettsf requested, if we want to protect this endpoint, the following lines should be added back to the `searchUsers` function:
`@Roles('admin') @UseGuards(AuthGuard('jwt'), RoleGuard)`


Also not sure what this shows, but if you alter the request to a misspelled role you receive the following return (if this input is controlled on the front end, we probably don't have to worry about it, I really don't know):
<img width="721" alt="Screenshot 2025-03-03 at 10 44 15 AM" src="https://github.com/user-attachments/assets/91e2193a-0ca7-4cf2-b318-d0063e493618" />

If you search without the role you get back all users, this maybe desired results:
<img width="721" alt="Screenshot 2025-03-03 at 10 44 42 AM" src="https://github.com/user-attachments/assets/221e9817-4a95-4e72-9552-1cda63aa42c3" />

bennettsf commented:
I think @CourtneyHoppus makes a good point with the misspelled role. 

Even though it shouldn't be possible to send a misspelled role to the backend with the dropdown options on the frontend, it might be best practice to have an ENUM or Object that contains all possible roles. If whatever role was sent to the backend isn't one of those valid roles, it should become an empty string (resulting in all roles being returned).

Asfand00 commented:
@bennettsf @CourtneyHoppus, I added back the `@Roles('admin') @UseGuards(AuthGuard('jwt'), RoleGuard)` as requested and also added an ENUM for handling the misspelled role issue you guys were talking about, let me know what you think of those small changes. Also if you prefer an exception being thrown instead of roles being returned upon an empty string I can change 

this line: filters.role = ''; // Defaults to fetching all users (user.controller.ts L49) 

to this line: throw new BadRequestException(`Invalid role: ${filters.role}`); 

Here is my test of postman with an misspelled role:
![image](https://github.com/user-attachments/assets/3cafda6b-11fe-4cbb-a6e8-b3874e29cb15)

it simply returns all users (results in an empty string)

bennettsf commented:
These changes seem fine to me. Looks like Gary already had an ENUM made for us which is great, glad we could put it to use. @CourtneyHoppus if you wouldn't mind, could you take a look at this and re-review your requested changes for Ozzy?

CourtneyHoppus reviewed (approved):
Tested as before, looks like everything is working as expected.
